### PR TITLE
refactor: Switch from JavaScript to TypeScript for auth pages

### PR DESCRIPTION
**Pull request by bennyn**
## Iterations:

- [x] Make auth pages compile with TypeScript
- [x] Type refinements
- [x] Replace `any` in `actions` with proper types
- [x] Replace `any` in `ActionCreator` with proper types
- [x] Replace `any` in `components` with proper types
- [x] Replace `any` in `pages` with proper types
- [x] Replace `any` in `reducer` with proper types
- [x] Enable `noImplicitAny` compiler flag
- [x] Use `@babel/preset-typescript` in favour of `awesome-typescript-loader` (boosts from ~15s to 11s build time)

### CODE REVIEW

File: app/script/auth/component/AccountForm.tsx [hunk 1]
```diff
@@ -230,13 +304,17 @@ class AccountForm extends PureComponent {
...
       isFetching: AuthSelector.isFetching(state),
       isPersonalFlow: AuthSelector.isPersonalFlow(state),
       isPersonalInvitationFlow: AuthSelector.isPersonalInvitationFlow(state),
     }),
-    {...AuthAction, ...UserAction}
+    (dispatch: ThunkDispatch): DispatchProps => ({
+      doSendActivationCode: (email: string) => dispatch(ROOT_ACTIONS.userAction.doSendActivationCode(email)),
+      pushAccountRegistrationData: (registrationData: Partial<RegistrationDataState>) =>
+        dispatch(ROOT_ACTIONS.authAction.pushAccountRegistrationData(registrationData)),
+    })
```
Comment (gregor) [inline app/script/auth/component/AccountForm.tsx]: Let's stick to explicit returns in {} functions if things break to multiline

File: app/script/auth/component/AppAlreadyOpen.tsx [hunk 2]
```diff
@@ -61,9 +73,11 @@ class AppAlreadyOpen extends React.Component {
...
 export default injectIntl(
   connect(
-    state => ({
+    (state: RootState): ConnectedProps => ({
       isAppAlreadyOpen: CookieSelector.isAppAlreadyOpen(state),
     }),
-    {...CookieAction}
+    (dispatch: ThunkDispatch): DispatchProps => ({
+      removeCookie: (name: string) => dispatch(ROOT_ACTIONS.cookieAction.removeCookie(name)),
+    })
```
Comment (gregor) [inline app/script/auth/component/AppAlreadyOpen.tsx]: Explicit return in multiline function. This is not easier to read.

File: app/script/auth/component/ClientItem.tsx [hunk 3]
```diff
@@ -112,16 +136,13 @@ class ClientItem extends React.Component {
...
     this.resetState();
     this.props.onClick(event);
   };
 
-  handleSubmit = event => {
+  handleSubmit = (event: React.FormEvent) => {
     event.preventDefault();
-    if (this.props.isFetching) {
-      return;
-    }
```
Comment (gregor) [inline app/script/auth/component/ClientItem.tsx:124]: Where do we handle this now?

File: app/script/auth/component/ClientList.tsx [hunk 4]
```diff
@@ -110,12 +129,19 @@ class ClientList extends React.Component {
...
       }),
-      {...ClientAction, ...LocalStorageAction, resetError}
+      (dispatch: ThunkDispatch): DispatchProps => ({
+        resetAuthError: () => dispatch(ROOT_ACTIONS.authAction.resetAuthError()),
+        doRemoveClient: (clientId: string, password: string) =>
+          dispatch(ROOT_ACTIONS.clientAction.doRemoveClient(clientId, password)),
+        getLocalStorage: (key: string) => dispatch(ROOT_ACTIONS.localStorageAction.getLocalStorage(key)),
+        doInitializeClient: (clientType: ClientType, password?: string) =>
+          dispatch(ROOT_ACTIONS.clientAction.doInitializeClient(clientType, password)),
+      })
```
Comment (gregor) [inline app/script/auth/component/ClientList.tsx]: Explicit return

File: app/script/auth/component/UnsupportedBrowser.tsx [hunk 5]
```diff
@@ -68,21 +85,31 @@ export const UnsupportedBrowser = ({
...
 
 export default injectIntl(
-  connect(state => ({
-    hasCookieSupport: RuntimeSelector.hasCookieSupport(state),
-    hasIndexedDbSupport: RuntimeSelector.hasIndexedDbSupport(state),
-    isCheckingSupport: RuntimeSelector.isChecking(state),
-    isSupportedBrowser: RuntimeSelector.isSupportedBrowser(state),
-  }))(UnsupportedBrowser)
+  connect(
+    (state: RootState): ConnectedProps => ({
```
Comment (gregor) [inline app/script/auth/component/UnsupportedBrowser.tsx]: explicit

File: app/script/auth/component/WirelessUnsupportedBrowser.tsx [hunk 6]
```diff
@@ -49,7 +62,10 @@ export const WirelessUnsupportedBrowser = ({children, isSupportedBrowser, intl:
   );
 
 export default injectIntl(
-  connect(state => ({
-    isSupportedBrowser: RuntimeSelector.isSupportedBrowser(state),
-  }))(WirelessUnsupportedBrowser)
+  connect(
+    (state: RootState): ConnectedProps => ({
+      isSupportedBrowser: RuntimeSelector.isSupportedBrowser(state),
+    }),
```
Comment (gregor) [inline app/script/auth/component/WirelessUnsupportedBrowser.tsx]: explicit

File: app/script/auth/component/WirelessUnsupportedBrowser.tsx [hunk 7]
```diff
@@ -49,7 +62,10 @@ export const WirelessUnsupportedBrowser = ({children, isSupportedBrowser, intl:
   );
```
Comment (gregor) [inline app/script/auth/component/WirelessUnsupportedBrowser.tsx:62]: explicit

File: app/script/auth/module/action/AuthAction.ts [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+          clientType = persist ? ClientType.PERMANENT : ClientType.TEMPORARY;
+          return apiClient.init(clientType);
+        })
+        .then(() => core.init())
+        .then(() => this.persistAuthData(clientType, core, dispatch, localStorageAction))
+        .then(() => {
+          if (options.shouldValidateLocalClient) {
+            return dispatch(authAction.validateLocalClient());
+          }
+          return undefined;
```
Comment (gregor) [inline app/script/auth/module/action/AuthAction.ts:303]: Ternary

File: app/script/auth/module/action/AuthAction.ts [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+      {apiClient, core, actions: {authAction, clientAction, selfAction, localStorageAction}}
+    ) => {
+      let clientType: ClientType;
+      dispatch(AuthActionCreator.startRefresh());
+      return Promise.resolve()
+        .then(() => {
+          if (options.isImmediateLogin) {
+            return dispatch(localStorageAction.setLocalStorage(LocalStorageKey.AUTH.PERSIST, true));
+          }
+          return undefined;
```
Comment (gregor) [inline app/script/auth/module/action/AuthAction.ts:287]: ternary

File: app/script/auth/module/action/ClientAction.ts [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+  doInitializeClient = (clientType: ClientType, password?: string): ThunkAction => {
+    return (dispatch, getState, {core, actions: {clientAction, notificationAction}}) => {
+      dispatch(ClientActionCreator.startInitializeClient());
+      return Promise.resolve()
+        .then(() => core.initClient({clientType, password}, clientAction.generateClientPayload(clientType)))
+        .then(creationStatus =>
+          Promise.resolve()
+            .then(() => dispatch(ClientActionCreator.successfulInitializeClient(creationStatus)))
+            .then(() => creationStatus)
+        )
```
Comment (gregor) [inline app/script/auth/module/action/ClientAction.ts:70]: {} and explicit return for statements that break the single line

File: app/script/auth/module/action/creator/AuthActionCreator.ts [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+    type: AUTH_ACTION.GET_INVITATION_FROM_CODE_START,
+  });
+  static successfulGetInvitationFromCode = (invitation: Invitation): GetInvitationFromCodeSuccessAction => ({
+    payload: invitation,
+    type: AUTH_ACTION.GET_INVITATION_FROM_CODE_SUCCESS,
+  });
+  static failedGetInvitationFromCode = (error: Error): GetInvitationFromCodeFailedAction => ({
+    error,
+    type: AUTH_ACTION.GET_INVITATION_FROM_CODE_FAILED,
+  });
```
Comment (gregor) [inline app/script/auth/module/action/creator/AuthActionCreator.ts]: lots of explicit returns

File: app/script/auth/module/action/creator/ClientActionCreator.ts [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+    type: CLIENT_ACTION.CLIENT_INIT_SUCCESS,
+  });
+  static failedInitializeClient = (error: Error): InitializeClientFailedAction => ({
+    error,
+    type: CLIENT_ACTION.CLIENT_INIT_FAILED,
+  });
+
+  static resetError = (): ResetClientErrorsAction => ({
+    type: CLIENT_ACTION.CLIENT_RESET_ERROR,
+  });
```
Comment (gregor) [inline app/script/auth/module/action/creator/ClientActionCreator.ts:136]: explizit returns

File: app/script/auth/module/action/creator/ConversationActionCreator.ts [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    type: CONVERSATION_ACTION.CONVERSATION_CODE_JOIN_START,
+  });
+  static successfulJoinConversationByCode = (data: ConversationEvent): ConversationCodeJoinSuccessAction => ({
+    payload: data,
+    type: CONVERSATION_ACTION.CONVERSATION_CODE_JOIN_SUCCESS,
+  });
+  static failedJoinConversationByCode = (error: Error): ConversationCodeJoinFailedAction => ({
+    error,
+    type: CONVERSATION_ACTION.CONVERSATION_CODE_JOIN_FAILED,
+  });
```
Comment (gregor) [inline app/script/auth/module/action/creator/ConversationActionCreator.ts:86]: explizit returns

File: app/script/auth/module/action/creator/CookieActionCreator.ts [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+    type: COOKIE_ACTION.COOKIE_REMOVE_START,
+  });
+  static successfulRemoveCookie = ({cookie, name}: {cookie?: string; name: string}): CookieRemoveSuccessAction => ({
+    payload: {cookie, name},
+    type: COOKIE_ACTION.COOKIE_REMOVE_SUCCESS,
+  });
+  static failedRemoveCookie = (error: Error): CookieRemoveFailedAction => ({
+    error,
+    type: COOKIE_ACTION.COOKIE_REMOVE_FAILED,
+  });
```
Comment (gregor) [inline app/script/auth/module/action/creator/CookieActionCreator.ts:157]: explizit returns

File: app/script/auth/module/action/creator/InvitationActionCreator.ts [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+    type: INVITATION_ACTION.INVITE_ADD_SUCCESS,
+  });
+  static failedAddInvite = (error: Error): AddInviteFailedAction => ({
+    error,
+    type: INVITATION_ACTION.INVITE_ADD_FAILED,
+  });
+
+  static resetError = (): ResetInviteErrorAction => ({
+    type: INVITATION_ACTION.INVITE_RESET_ERROR,
+  });
```
Comment (gregor) [inline app/script/auth/module/action/creator/InvitationActionCreator.ts:67]: explizit returns

File: app/script/auth/module/action/creator/LanguageActionCreator.ts [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    type: LANGUAGE_ACTION.SWITCH_LANGUAGE_START,
+  });
+  static successfulSwitchLanguage = (language: string): LanguageSwitchSuccessAction => ({
+    payload: language,
+    type: LANGUAGE_ACTION.SWITCH_LANGUAGE_SUCCESS,
+  });
+  static failedSwitchLanguage = (error: Error): LanguageSwitchFailedAction => ({
+    error,
+    type: LANGUAGE_ACTION.SWITCH_LANGUAGE_FAILED,
+  });
```
Comment (gregor) [inline app/script/auth/module/action/creator/LanguageActionCreator.ts:53]: explizit returns

File: app/script/auth/module/action/creator/LocalStorageActionCreator.ts [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+    type: LOCAL_STORAGE_ACTION.LOCAL_STORAGE_DELETE_START,
+  });
+  static successfulLocalStorageDelete = (key: string): DeleteSuccessAction => ({
+    payload: {key},
+    type: LOCAL_STORAGE_ACTION.LOCAL_STORAGE_DELETE_SUCCESS,
+  });
+  static failedLocalStorageDelete = (error: Error): DeleteFailedAction => ({
+    error,
+    type: LOCAL_STORAGE_ACTION.LOCAL_STORAGE_DELETE_FAILED,
+  });
```
Comment (gregor) [inline app/script/auth/module/action/creator/LocalStorageActionCreator.ts:118]: explizit returns

File: app/script/auth/module/action/creator/NotificationActionCreator.ts [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+    type: NOTIFICATION_ACTION.NOTIFICATION_CHECK_HISTORY_SUCCESS,
+  });
+  static failedCheckHistory = (error: Error): HistoryCheckFailedAction => ({
+    error,
+    type: NOTIFICATION_ACTION.NOTIFICATION_CHECK_HISTORY_FAILED,
+  });
+
+  static resetHistoryCheck = (): HistoryCheckResetAction => ({
+    type: NOTIFICATION_ACTION.NOTIFICATION_CHECK_HISTORY_RESET,
+  });
```
Comment (gregor) [inline app/script/auth/module/action/creator/NotificationActionCreator.ts:67]: explizit returns

File: app/script/auth/module/action/creator/RuntimeActionCreator.ts [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+    type: RUNTIME_ACTION.RUNTIME_CHECK_COOKIE_START,
+  });
+  static finishCheckCookie = (result: boolean): CheckCookieSuccessAction => ({
+    payload: result,
+    type: RUNTIME_ACTION.RUNTIME_CHECK_COOKIE_FINISH,
+  });
+
+  static confirmSupportedBrowser = (): ConfirmSupportedBrowserAction => ({
+    type: RUNTIME_ACTION.RUNTIME_CONFIRM_SUPPORTED_BROWSER,
+  });
```
Comment (gregor) [inline app/script/auth/module/action/creator/RuntimeActionCreator.ts:76]: explizit returns

File: app/script/auth/module/action/creator/SelfActionCreator.ts [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+    type: SELF_ACTION.CONSENT_SET_START,
+  });
+  static successfulSetConsent = (consent: Consent): SetConsentSuccessAction => ({
+    payload: consent,
+    type: SELF_ACTION.CONSENT_SET_SUCCESS,
+  });
+  static failedSetConsent = (error: Error): SetConsentFailedAction => ({
+    error,
+    type: SELF_ACTION.CONSENT_SET_FAILED,
+  });
```
Comment (gregor) [inline app/script/auth/module/action/creator/SelfActionCreator.ts:150]: explizit returns

File: app/script/auth/module/action/creator/UserActionCreator.ts [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+  static startSendActivationCode = (): SendActivationCodeStartAction => ({
+    type: USER_ACTION.USER_SEND_ACTIVATION_CODE_START,
+  });
+  static successfulSendActivationCode = (): SendActivationCodeSuccessAction => ({
+    type: USER_ACTION.USER_SEND_ACTIVATION_CODE_SUCCESS,
+  });
+  static failedSendActivationCode = (error: Error): SendActivationCodeFailedAction => ({
+    error,
+    type: USER_ACTION.USER_SEND_ACTIVATION_CODE_FAILED,
+  });
```
Comment (gregor) [inline app/script/auth/module/action/creator/UserActionCreator.ts:54]: explizit returns

File: app/script/auth/module/selector/AuthSelector.ts [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+export const getError = (state: RootState) => state.authState.error;
+export const getCurrentFlow = (state: RootState) => state.authState.currentFlow;
+export const getAccount = (state: RootState): RegistrationDataState => state.authState.account || unsetRegistrationData;
+export const getAccountTeam = (state: RootState) => getAccount(state).team || unsetTeam;
+export const getAccountTeamName = (state: RootState) => getAccountTeam(state).name;
+
+export const isGenericInvitationFlow = (state: RootState) => getCurrentFlow(state) === REGISTER_FLOW.GENERIC_INVITATION;
+export const isPersonalFlow = (state: RootState) => getCurrentFlow(state) === REGISTER_FLOW.PERSONAL;
+export const isPersonalInvitationFlow = (state: RootState) =>
+  getCurrentFlow(state) === REGISTER_FLOW.PERSONAL_INVITATION;
```
Comment (gregor) [inline app/script/auth/module/selector/AuthSelector.ts:66]: explizit return

File: app/script/auth/module/selector/RuntimeSelector.ts [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+*
+* You should have received a copy of the GNU General Public License
+* along with this program. If not, see http://www.gnu.org/licenses/.
+*
+*/
+
+import {RootState} from '../reducer';
+
+export const isChecking = (state: RootState) =>
+  state.runtimeState.isCheckingIndexedDb || state.runtimeState.isCheckingCookie;
```
Comment (gregor) [inline app/script/auth/module/selector/RuntimeSelector.ts:23]: explicit

File: app/script/auth/page/ChooseHandle.tsx [hunk 24]
```diff
@@ -133,21 +152,22 @@ class ChooseHandle extends React.PureComponent {
...
-        doSetConsent,
-        setHandle,
-      }
+      (dispatch: ThunkDispatch): DispatchProps => ({
+        checkHandles: (handles: string[]) => dispatch(ROOT_ACTIONS.userAction.checkHandles(handles)),
+        doGetConsents: () => dispatch(ROOT_ACTIONS.selfAction.doGetConsents()),
+        doSetConsent: (consentType: ConsentType, value: number) =>
+          dispatch(ROOT_ACTIONS.selfAction.doSetConsent(consentType, value)),
+        setHandle: (handle: string) => dispatch(ROOT_ACTIONS.selfAction.setHandle(handle)),
+      })
```
Comment (gregor) [inline app/script/auth/page/ChooseHandle.tsx]: explicit

File: app/script/auth/component/ClientItem.tsx (hunk 3 above)
Comment (bennyn) [reply app/script/auth/component/ClientItem.tsx]: I just copied over the functionality that was already present: https://github.com/wireapp/wire-webapp/blob/2018-10-02-08-03/app/script/auth/component/ClientItem.jsx#L122-L124

Comment (Yserz) [reply app/script/auth/component/ClientItem.tsx:124]: This was a false friend. We never handled this state so it's better to remove it.

**Pull request merged**